flepimop/gempyor_pkg/src/gempyor/sync/_sync.py

```diff
@@ -1,336 +1,528 @@
-                [
-                    "echo",
-                    "`{}` failed with return code {}".format(
-                        " ".join(res.args), res.returncode
-                    ),
-                ],
-                stdout=res.stdout,
-                stderr=res.stderr,
-            )
-        else:
-            return res
-    except FileNotFoundError as e:
-        return run(["echo", "command `{}` not found".format(cmd[0])])
+from ..logging import get_script_logger
+from ..utils import _shutil_which
+from ._sync_filter import FilterParts, ListSyncFilter, WithFilters
+
+
+_RSYNC_HOST_REGEX: Final[Pattern[str]] = re.compile(r"^(?P<host>[^:]+):(?P<path>.+)$")
 
 
 class SyncOptions(BaseModel):
     """
-    The potential overriding options for a sync operation
-    :param target_override: optional: override the sync target
-    :param source_override: optional: override the sync source
-    :param filter_override: optional: override the sync filters; n.b. this is strict override, not an append/prepend
-    :param dryrun: optional (default: false) perform a dry run of the sync operation
-    :param reverse: optional (default: false) perform the sync operation in reverse (e.g. swap source and target)
+    Override options for sync protocols.
+
+    Attributes:
+        protocol: The sync protocol to override if not `None`.
+        source_override: Override for the sync source if not `None`.
+        target_override: Override for the sync target if not `None`.
+        source_append: Append to the source path if not `None`.
+        target_append: Append to the target path if not `None`.
+        filter_override: Override the sync filters. This is strict override, not an
+            append/prepend.
+        filter_prefix: Filters to prepend to the filter list.
+        filter_suffix: Filters to append to the filter list.
+        dry_run: If `True` perform a dry run of the sync operation, otherwise perform
+            the sync operation.
+        reverse: If `True` perform the sync operation in reverse (swap source and
+            target).
+        mkpath: If `True` create the target directory if it does not exist.
     """
 
     protocol: str | None = None
-    target_override: Path | None = None
-    source_override: Path | None = None
+    source_override: str | None = None
+    target_override: str | None = None
+    source_append: str | None = None
+    target_append: str | None = None
     filter_override: ListSyncFilter | None = None
     filter_prefix: ListSyncFilter = []
     filter_suffix: ListSyncFilter = []
     # n.b. a filter_override = [] would be a valid override, e.g. to clear filters
 
-    dryrun: bool = False
+    dry_run: bool = False
     reverse: bool = False
     mkpath: bool = False
+    sep: str = "/"
 
     # allow potentially other fields for external modules
     model_config = ConfigDict(extra="allow")
 
+    def _true_path(self, original: str, override: str | None, append: str | None) -> str:
```

Comment:
nice one.

this approach would also simplify the override vs append logic in previous comment.